---
title: "Methylation - EDA"
format: html
---

### Libraries

```{r}

# Plotting and graphics
library(ggplot2)
library(RColorBrewer)
library(scales)
library(colorspace)
library(cowplot)
library(npreg)
library(ComplexHeatmap)
library(patchwork)
library(circlize)

# Data QC 
library(wateRmelon)
library(DNAmArray)
library(minfi)
library(sva)
library(maxprobes)

# Annotation
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(org.Hs.eg.db)
library(AnnotationDbi)

# Programming, parsing and data wrangling
library(mgsub)
library(GenomicFeatures)
#library(doSNOW) # parallel processes on Windows
library(foreach)
library(RcppEigen) # For faster lm 

# Statistics and analyses
library(limma)
library(factoextra)
library(sva)
library(pvca)
library(dmrff)
library(DMRcate)
library(variancePartition)

# Gene ontology
library(clusterProfiler)
#library(methylGSA)

```

### Parameters

```{r}
#set the treat log2 treshold
trt_log2fc_thr <- 0.1
propProbes_varPart <- 0.1

```

### Directories

```{r}

# Defined main dir
scripts_dir <- getwd()
main_dir <- "methylation/"
data_dir <- "rawdata/"
res_dir <- "results/"
pca_dir <- "pca/"
dmr_dir <- "dmr/"
de_dir <- "differential_methylation/"
util_dir <- "utilities/"

```

## Read in the pre-processed data

```{r}
preProccessedData <- readRDS(paste0(data_dir,"preproccessed_methyl.RDS"))
qnt.mSetSqFlt <- preProccessedData$qnt.mSetSqFlt
fun.mSetSqFlt <- preProccessedData$fun.mSetSqFlt
metadata <- preProccessedData$metadata
annotdata <- preProccessedData$annotdata
```


```{r}

# Categorial
sex_pal <- c("green", "mediumpurple")
array_pal <- hue_pal()(length(unique(metadata$Array)))
well_pal <- brewer.pal(length(unique(metadata$ARRAY_ROW)),"Dark2")
batch_pal <- brewer.pal(length(unique(metadata$Batch)),"Set1"); names(batch_pal) <- unique(metadata$Batch) 

# Define BMI group colours
bmi_cols <- npreg::number2color(c(1,2,3,4,5), c("yellow","darkred"), equidistant = FALSE, xmin=1, xmax=5)
names(bmi_cols) <- c("Underweight","Normal","Overweight","Obese","Extremely Obese")

# Continuous
weeks.col_ramp <- colour_ramp(c("white","darkgreen"), na.color="grey", alpha=TRUE)
mat_age.col_ramp <- colour_ramp(c("white","blue"), na.color="grey", alpha=TRUE)
bmi.col_ramp <- colour_ramp(c("white","darkorange"), na.color="grey", alpha=TRUE)

```

#### Correlate PCs with continuous metadata

```{r}

# Read in expanded metadata
expandedMeta <- readxl::read_xlsx("Additional-MEtadata.xlsx", sheet=1)
table(expandedMeta$`Processing Date`)

# Fix variables
expandedMeta["PROC_DATE"] <- as.character(as.numeric(expandedMeta$`Processing Date`))

# Combine with PCA metadata
metadata <- merge(metadata, expandedMeta[,c("Embryo ID","Tissue (mg)","PROC_DATE", "DNA (ng/ul)", "Total DNA", "DNA (ng/mg)", "260/280")], by.x="Sample_Name", by.y="Embryo ID")

```


#### Pre-process metadata 

```{r}

# Scale covariates
metadata["Batch"] <- paste0("B_",metadata$Batch)
batch_pal <- brewer.pal(length(unique(metadata$Batch)),"Set1"); names(batch_pal) <- unique(metadata$Batch) 

metadata["Z_DNAul"] <- scale(metadata$`DNA (ng/ul)`, center=TRUE, scale=TRUE)
metadata["Z_DNAmg"] <- scale(metadata$`DNA (ng/mg)`, center=TRUE, scale=TRUE)
metadata["Z_TISSUEmg"] <- scale(metadata$`Tissue (mg)`, center=TRUE, scale=TRUE)
metadata["Z_RRNA_ABSORB"] <- scale(metadata$`260/280`, center=TRUE, scale=TRUE)

# Scale data
metadata["Z_STAGE_WEEKS"] <- scale(metadata$stage_weeks, scale=TRUE, center=TRUE)
metadata["Z_MAT_BMI"] <- scale(metadata$mat_bmi, scale=TRUE, center=TRUE)
metadata["Z_MAT_AGE"] <- scale(metadata$mat_age, scale=TRUE, center=TRUE)
rownames(metadata) <- metadata$Sample_Name

# Fix id names
metadata$Sample_Name <- as.character(metadata$Sample_Name)
rownames(metadata) <- metadata$Sample_Name

```

#### Get ordered M-values

```{r}

# Get M-normalised matrix
## Quantile
qnt.m_mat <- minfi::getM(qnt.mSetSqFlt)
qnt.m_mat <- as.matrix(na.omit(qnt.m_mat))
qnt.m_mat <- qnt.m_mat[order(rowVars(qnt.m_mat), decreasing=TRUE),]

## Functional
fun.m_mat <- minfi::getM(fun.mSetSqFlt)
fun.m_mat <- as.matrix(na.omit(fun.m_mat))
fun.m_mat <- fun.m_mat[order(rowVars(fun.m_mat), decreasing=TRUE),]

```

#### SVA

No surrogate variables after taking into account known batch. Currently errors with `Error in solve.default(t(mod0) %*% mod0) : 'a' is 0-diml`.

```{r}

# Define matrices
set.seed(42)

# Drop probes with problem values
dim(fun.m_mat)
fun.m_mat <- fun.m_mat[apply(fun.m_mat, 1, function(row) all(!is.na(row)) & all(is.finite(row))),]
dim(fun.m_mat)

# Define models
modBatch <- model.matrix(~ 0 + sex + Batch + Z_STAGE_WEEKS + Z_MAT_BMI + Z_MAT_AGE, data=metadata) 
modBatchBig <- model.matrix(~ 0 + sex + Batch + Z_STAGE_WEEKS + Z_MAT_BMI + Z_MAT_AGE + Z_DNAul +  Z_TISSUEmg + Z_RRNA_ABSORB, data=metadata) 

# Run SV analysis
nSVBatch <- num.sv(fun.m_mat[,rownames(modBatch)], modBatch, method="leek"); nSVBatch 
nSVBatchBig <- num.sv(fun.m_mat[,rownames(modBatchBig)], modBatchBig, method="leek"); nSVBatchBig

```



#### Combat

This checks how equivalent array row and sentrix id (batch) are and which is best to adjust for.


```{r}

# Define combat matrix
modCombat <- model.matrix(~ sex + Z_MAT_BMI + Z_STAGE_WEEKS + Z_MAT_AGE + Z_RRNA_ABSORB + Z_DNAul +  Z_TISSUEmg, data=metadata)

# Adjust
metadata$Sample_Name <- as.character(metadata$Sample_Name)
combatBatchAdj <- ComBat(dat=fun.m_mat[,metadata$Sample_Name], batch=metadata$Batch, mod=modCombat[rownames(metadata),], par.prior=TRUE, prior.plots=TRUE)
combatRowAdj <- ComBat(dat=fun.m_mat[,metadata$Sample_Name], batch=metadata$ARRAY_ROW, mod=modCombat[rownames(metadata),], par.prior=TRUE, prior.plots=TRUE)

```

#### Fit variance partition models

Note that when including a random slope \[ie (XXX + 0\|YYY)\], total variance won't sum to 1. So % bar plots aren't applicable.

```{r eval=FALSE}
set.seed(42)

# Define formula 
varPartFormula <- ~ (1|sex) + (1|Batch) + Z_MAT_BMI + Z_STAGE_WEEKS + Z_MAT_AGE + Z_RRNA_ABSORB + Z_DNAul + Z_TISSUEmg


metadata$Batch <- paste0("B_",metadata$Batch)
metadata$PROC_DATE <- paste0("B_",metadata$PROC_DATE)
# Fit models
varPart <- fitExtractVarPartModel(fun.m_mat[sample(1:nrow(fun.m_mat), floor(propProbes_varPart*nrow(fun.m_mat))),], varPartFormula, metadata)
# varPartBatch <- fitExtractVarPartModel(combatBatchAdj[sample(1:nrow(combatBatchAdj), floor(propProbes_varPart*nrow(combatBatchAdj))),metadata$Sample_Name], varPartFormula, metadata)
# varPartRow <- fitExtractVarPartModel(combatRowAdj[sample(1:nrow(combatRowAdj), floor(propProbes_varPart*nrow(combatRowAdj))),metadata$Sample_Name], varPartFormula, metadata)
saveRDS(varPart, paste0(pca_dir, "varPartition_model"))
# saveRDS(varPartBatch, paste0(pca_dir, "varPartition_batchAdjust_model"))




# Plot
varPartPlt <- sortCols(varPart) %>% plotVarPart() +
  cowplot::theme_cowplot(font_size = 16) + 
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90)) +
  ggtitle("Variance Partition Analysis: Un-Adjusted")

# 
# varPartBatchPlt <- plotVarPart(varPartBatch) + 
#                           theme_bw(base_size=24) + 
#                           theme(axis.title=element_text(size=42),
#                                 axis.text.x=element_text(size=32, angle=90, hjust=1, vjust=0.5),
#                                 legend.text=element_text(size=16),
#                                 legend.title=element_blank()) + 
#                           ggtitle("Variance Partition Analysis: Sentrix ID Batch-Adjusted")
# varPartRowPlt <- plotVarPart(varPartBatch) + 
#                           theme_bw(base_size=24) + 
#                           theme(axis.title=element_text(size=42),
#                                 axis.text.x=element_text(size=32, angle=90, hjust=1, vjust=0.5),
#                                 legend.text=element_text(size=16),
#                                 legend.title=element_blank()) + 
#                           ggtitle("Variance Partition Analysis: Sentrix Posiiton Batch-Adjusted")

# Plot
ggsave(paste0(pca_dir, "foetalSamples_varPartitionPlot.png"), varPartPlt, width=7, height=5, units="in")
# ggsave(paste0(pca_dir, "foetalSamples_sentrixIDAdjusted_varPartitionPlot.png"), varPartBatchPlt, width=6000, height=5000, units="px")
# ggsave(paste0(pca_dir, "foetalSamples_sentrixPositionAdjusted_varPartitionPlot.png"), varPartBatchPlt, width=6000, height=5000, units="px")

```

#### Manual PCA

```{r}

source(paste0(util_dir,"R/pca_utility_functions.R"))

prop_probes.by_var <- 0.95

# Run PCA
qnt.pca_res <- prcomp(t(qnt.m_mat[1:(nrow(qnt.m_mat)*prop_probes.by_var),]), center=TRUE, scale=TRUE)
fun.pca_res <- prcomp(t(fun.m_mat[1:(nrow(fun.m_mat)*prop_probes.by_var),]), center=TRUE, scale=TRUE)
bat.pca_res <- prcomp(t(combatBatchAdj[1:(nrow(combatBatchAdj)*prop_probes.by_var),]), center=TRUE, scale=TRUE)
row.pca_res <- prcomp(t(combatRowAdj[1:(nrow(combatRowAdj)*prop_probes.by_var),]), center=TRUE, scale=TRUE)

# Merge with metadata
metadata$BMI_GROUP <- factor(metadata$BMI_GROUP, levels=c("Underweight","Normal","Overweight","Obese","Extremely Obese"))
qnt.pca_metadata <- merge(metadata[,!grepl("^PC",colnames(metadata))], qnt.pca_res$x, by.x="Sample_Name", by.y="row.names")
fun.pca_metadata <- merge(metadata[,!grepl("^PC",colnames(metadata))], fun.pca_res$x, by.x="Sample_Name", by.y="row.names")
bat.pca_metadata <- merge(metadata[,!grepl("^PC",colnames(metadata))], bat.pca_res$x, by.x="Sample_Name", by.y="row.names")
row.pca_metadata <- merge(metadata[,!grepl("^PC",colnames(metadata))], row.pca_res$x, by.x="Sample_Name", by.y="row.names")

```

#### Plot Functional Normalisation PCA

```{r}

# Reorder sex match
fun.pca_metadata$SEX_MATCH <- factor(fun.pca_metadata$SEX_MATCH, levels=c("DIFFERENT","SAME"))

# Loop over consecutive PCs
for (pc in list(c(1,2), c(2,3), c(3,4), c(4,5))) {

  # Generate plots
  sex_plt <- Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="sex", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=c("green","mediumpurple"))
  row_plt <- Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="ARRAY_ROW", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=well_pal)
  bmig_plt <- Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="BMI_GROUP", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=bmi_cols)
  batch_plt <- Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="Batch", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=batch_pal)
  age_plt <- Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="mat_age", label_id="Sample_Name", pcs=pc) + scale_fill_gradient(high="darkblue", low="white")
  weeks_plt <- Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="stage_weeks", label_id="Sample_Name", pcs=pc) + scale_fill_gradient(high="darkgreen", low="white")
  tri_plt <- Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="TRIMESTER", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=c("#FEE391", "#EC7014", "#662506"))
  bmi_plt <- Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="mat_bmi", label_id="Sample_Name", pcs=pc) + scale_fill_gradient(high="darkorange", low="white")
  proc_plt <- Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="PROC_DATE", label_id="Sample_Name", pcs=pc)
  
  # Save
  dir.create(paste0(pca_dir, "unadjusted/"), recursive=TRUE)
  ggsave(paste0(pca_dir, "unadjusted/pca_biplot.fun_norm.pc",paste(pc, collapse=""),".colby_sex.propprobes",prop_probes.by_var,".png"), sex_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "unadjusted/pca_biplot.fun_norm.pc",paste(pc, collapse=""),".colby_arrayrow.propprobes",prop_probes.by_var,".png"), row_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "unadjusted/pca_biplot.fun_norm.pc",paste(pc, collapse=""),".colby_bmigroup.propprobes",prop_probes.by_var,".png"), bmig_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "unadjusted/pca_biplot.fun_norm.pc",paste(pc, collapse=""),".colby_batch.propprobes",prop_probes.by_var,".png"), batch_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "unadjusted/pca_biplot.fun_norm.pc",paste(pc, collapse=""),".colby_maternalage.propprobes",prop_probes.by_var,".png"), age_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "unadjusted/pca_biplot.fun_norm.pc",paste(pc, collapse=""),".colby_stageweeks.propprobes",prop_probes.by_var,".png"), weeks_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "unadjusted/pca_biplot.fun_norm.pc",paste(pc, collapse=""),".colby_trimester.propprobes",prop_probes.by_var,".png"), tri_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "unadjusted/pca_biplot.fun_norm.pc",paste(pc, collapse=""),".colby_bmi.propprobes",prop_probes.by_var,".png"), bmi_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "unadjusted/pca_biplot.fun_norm.pc",paste(pc, collapse=""),".colby_processDate.propprobes",prop_probes.by_var,".png"), proc_plt, units="px", width=3000, height=3000)
  
}


```

#### Compare functional vs quantile normalisation
```{r}
weeks_plt_qnt <- Plot2DPCA(meta=qnt.pca_metadata, var=get_eigenvalue(qnt.pca_res), color_by="stage_weeks", label_id="Sample_Name", pcs=1:2) + scale_fill_gradient(high="darkgreen", low="white") +theme(legend.position = "right") + ggtitle("Quantile normalisation - Stage")
weeks_plt_fun <- Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="stage_weeks", label_id="Sample_Name", pcs=1:2) + scale_fill_gradient(high="darkgreen", low="white") +theme(legend.position = "right") + ggtitle("Functional normalisation - Stage")

batch_plt_qnt <- Plot2DPCA(meta=qnt.pca_metadata, var=get_eigenvalue(qnt.pca_res), color_by="Batch", label_id="Sample_Name", pcs=1:2) + scale_fill_manual(values=batch_pal) +theme(legend.position = "right") + ggtitle("Quantile normalisation - Batch")
batch_plt_fun <- Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="Batch", label_id="Sample_Name", pcs=1:2) + scale_fill_manual(values=batch_pal)+theme(legend.position = "right") + ggtitle("Functional normalisation - Batch")

pcas <- (weeks_plt_qnt + weeks_plt_fun)/(batch_plt_qnt+batch_plt_fun) +  patchwork::plot_annotation(tag_levels = 'A')

ggsave(paste0(pca_dir, "unadjusted/pca_comparisons.png"), pcas, units="in", width=12, height=10)
ggsave("figures/SuppFig_pca_comparisons.pdf", pcas, units="in", width=12, height=10)



weeks_plt_fun <- Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="stage_weeks", label_id="Sample_Name", pcs=1:2) + scale_fill_gradient(high="darkgreen", low="white") +theme(legend.position = "right") + labs(fill="Stage\n(Weeks)")

ggsave(paste0(pca_dir, "unadjusted/Fig1A_pca_stage.png"), weeks_plt_fun, units="in", width=7, height=5)
ggsave("figures/Fig1A_pca_stage.pdf", weeks_plt_fun, units="in", width=7, height=5)

```
Correlation of covariates to the PCs
```{r}

covariates <- metadata[,c("sex","Batch","mat_bmi","stage_weeks","mat_age","260/280","DNA (ng/ul)","Tissue (mg)")]
covariates$Batch <- as.numeric(as.factor(covariates$Batch))
covariates$sex <- as.numeric(as.factor(covariates$sex))

#significance

n <- t(!is.na(covariates)) %*% (!is.na(fun.pca_metadata[,paste0("PC",1:15)]))
r <- cor(covariates,fun.pca_metadata[,paste0("PC",1:15)]) 

cor2pvalue = function(r, n) {
  t <- (r*sqrt(n-2))/sqrt(1-r^2)
  p <- 2*(1 - pt(abs(t),(n-2)))
  se <- sqrt((1-r*r)/(n-2))
  out <- list(r, n, t, p, se)
  names(out) <- c("r", "n", "t", "p", "se")
  return(out)
}
# get a list with matrices of correlation, pvalues, standard error, etc.
corSig <- cor2pvalue(r,n)$p


col_fun = colorRamp2(c(-1, 0, 1), c("lightblue", "white", "red"))

pdf("figures/SuppFig_pca_correlation.pdf", width=8, height=7)
Heatmap(r,cluster_columns=FALSE, name= "Pearson\nCorrelation", col = col_fun, cell_fun = function(j, i, x, y, w, h, fill) {
    if(corSig[i, j] < 0.001) {
        grid.text("***", x, y,gp = gpar(fontsize = 18))
    } else if(corSig[i, j] < 0.01) {
        grid.text("**", x, y,gp = gpar(fontsize = 18))
    } else if(corSig[i, j] < 0.05) {
        grid.text("*", x, y,gp = gpar(fontsize = 18))
    }
  })
dev.off()

```





#### Plotting of batch effect (sentrix ID) in the four PCA plots

```{r}

#loop through the PC combinations
pcs <- list(c(1,2), c(2,3), c(3,4), c(4,5))
plots <- lapply(pcs,function(pc) Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="Batch", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=batch_pal) +  cowplot::theme_cowplot(font_size = 18))

#make one plot
batchPlot <- plot_grid(plotlist = plots,labels=LETTERS[1:4])

#save the combined plot
ggsave(paste0(pca_dir, "unadjusted/batchEffectPCA.png"), batchPlot, units="in", width=12, height=7,bg = "white")
ggsave("figures/SuppFig_batchEffectPCA.pdf", batchPlot, units="in", width=12, height=7,bg = "white")

```

#### Plotting of stage (sentrix ID) in the four PCA plots

```{r}

#loop through the PC combinations
pcs <- list(c(1,2), c(2,3), c(3,4), c(4,5))
plots <- lapply(pcs,function(pc) Plot2DPCA(meta=fun.pca_metadata, var=get_eigenvalue(fun.pca_res), color_by="stage_weeks", label_id="Sample_Name", pcs=pc) +  scale_fill_gradient(high="darkgreen", low="white") +  cowplot::theme_cowplot(font_size = 18))

#make one plot
stagePlot <- plot_grid(plotlist = plots,labels=LETTERS[1:4])

#save the combined plot
ggsave(paste0(pca_dir, "unadjusted/stageEffectPCA.png"), stagePlot, units="in", width=12, height=7,bg = "white")

```


#### Plot Combat Sentrix ID-adjusted

```{r}

# Reorder sex match
bat.pca_metadata$SEX_MATCH <- factor(bat.pca_metadata$SEX_MATCH, levels=c("DIFFERENT","SAME"))

# Loop over consecutive PCs
for (pc in list(c(1,2), c(2,3), c(3,4), c(4,5))) {

  # Generate plots
  sex_plt <- Plot2DPCA(meta=bat.pca_metadata, var=get_eigenvalue(bat.pca_res), color_by="sex", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=c("green","mediumpurple"))
  row_plt <- Plot2DPCA(meta=bat.pca_metadata, var=get_eigenvalue(bat.pca_res), color_by="ARRAY_ROW", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=well_pal)
  bmig_plt <- Plot2DPCA(meta=bat.pca_metadata, var=get_eigenvalue(bat.pca_res), color_by="BMI_GROUP", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=bmi_cols)
  batch_plt <- Plot2DPCA(meta=bat.pca_metadata, var=get_eigenvalue(bat.pca_res), color_by="Batch", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=batch_pal)
  age_plt <- Plot2DPCA(meta=bat.pca_metadata, var=get_eigenvalue(bat.pca_res), color_by="mat_age", label_id="Sample_Name", pcs=pc) + scale_fill_gradient(high="darkblue", low="white")
  weeks_plt <- Plot2DPCA(meta=bat.pca_metadata, var=get_eigenvalue(bat.pca_res), color_by="stage_weeks", label_id="Sample_Name", pcs=pc) + scale_fill_gradient(high="darkgreen", low="white")
  tri_plt <- Plot2DPCA(meta=bat.pca_metadata, var=get_eigenvalue(bat.pca_res), color_by="TRIMESTER", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=c("#FEE391", "#EC7014", "#662506"))
  bmi_plt <- Plot2DPCA(meta=bat.pca_metadata, var=get_eigenvalue(bat.pca_res), color_by="mat_bmi", label_id="Sample_Name", pcs=pc) + scale_fill_gradient(high="darkorange", low="white")
  proc_plt <- Plot2DPCA(meta=bat.pca_metadata, var=get_eigenvalue(bat.pca_res), color_by="PROC_DATE", label_id="Sample_Name", pcs=pc)
  
  # Save
  dir.create(paste0(pca_dir, "comBat_sentrixID/"), recursive=TRUE)
  ggsave(paste0(pca_dir, "comBat_sentrixID/pca_biplot.qnt_norm.combatBatchAdj.pc",paste(pc, collapse=""),".colby_sex.propprobes",prop_probes.by_var,".png"), sex_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixID/pca_biplot.qnt_norm.combatBatchAdj.pc",paste(pc, collapse=""),".colby_arrayrow.propprobes",prop_probes.by_var,".png"), row_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixID/pca_biplot.qnt_norm.combatBatchAdj.pc",paste(pc, collapse=""),".colby_bmigroup.propprobes",prop_probes.by_var,".png"), bmig_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixID/pca_biplot.qnt_norm.combatBatchAdj.pc",paste(pc, collapse=""),".colby_batch.propprobes",prop_probes.by_var,".png"), batch_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixID/pca_biplot.qnt_norm.combatBatchAdj.pc",paste(pc, collapse=""),".colby_maternalage.propprobes",prop_probes.by_var,".png"), age_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixID/pca_biplot.qnt_norm.combatBatchAdj.pc",paste(pc, collapse=""),".colby_stageweeks.propprobes",prop_probes.by_var,".png"), weeks_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixID/pca_biplot.qnt_norm.combatBatchAdj.pc",paste(pc, collapse=""),".colby_trimester.propprobes",prop_probes.by_var,".png"), tri_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixID/pca_biplot.qnt_norm.combatBatchAdj.pc",paste(pc, collapse=""),".colby_bmi.propprobes",prop_probes.by_var,".png"), bmi_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixID/pca_biplot.qnt_norm.combatBatchAdj.pc",paste(pc, collapse=""),".colby_processDate.propprobes",prop_probes.by_var,".png"), proc_plt, units="px", width=3000, height=3000)
  
}

```

#### Plot Combat Sentrix Position-adjusted

```{r}

# Reorder sex match
row.pca_metadata$SEX_MATCH <- factor(row.pca_metadata$SEX_MATCH, levels=c("DIFFERENT","SAME"))

# Loop over consecutive PCs
for (pc in list(c(1,2), c(2,3), c(3,4), c(4,5))) {

  # Generate plots
  sex_plt <- Plot2DPCA(meta=row.pca_metadata, var=get_eigenvalue(row.pca_res), color_by="sex", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=c("green","mediumpurple"))
  row_plt <- Plot2DPCA(meta=row.pca_metadata, var=get_eigenvalue(row.pca_res), color_by="ARRAY_ROW", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=well_pal)
  bmig_plt <- Plot2DPCA(meta=row.pca_metadata, var=get_eigenvalue(row.pca_res), color_by="BMI_GROUP", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=bmi_cols)
  batch_plt <- Plot2DPCA(meta=row.pca_metadata, var=get_eigenvalue(row.pca_res), color_by="Batch", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=batch_pal)
  age_plt <- Plot2DPCA(meta=row.pca_metadata, var=get_eigenvalue(row.pca_res), color_by="mat_age", label_id="Sample_Name", pcs=pc) + scale_fill_gradient(high="darkblue", low="white")
  weeks_plt <- Plot2DPCA(meta=row.pca_metadata, var=get_eigenvalue(row.pca_res), color_by="stage_weeks", label_id="Sample_Name", pcs=pc) + scale_fill_gradient(high="darkgreen", low="white")
  tri_plt <- Plot2DPCA(meta=row.pca_metadata, var=get_eigenvalue(row.pca_res), color_by="TRIMESTER", label_id="Sample_Name", pcs=pc) + scale_fill_manual(values=c("#FEE391", "#EC7014", "#662506"))
  bmi_plt <- Plot2DPCA(meta=row.pca_metadata, var=get_eigenvalue(row.pca_res), color_by="mat_bmi", label_id="Sample_Name", pcs=pc) + scale_fill_gradient(high="darkorange", low="white")
  
  # Save
  dir.create(paste0(pca_dir, "comBat_sentrixPosition/"), recursive=TRUE)
  ggsave(paste0(pca_dir, "comBat_sentrixPosition/pca_biplot.qnt_norm.combatRowAdj.pc",paste(pc, collapse=""),".colby_sex.propprobes",prop_probes.by_var,".png"), sex_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixPosition/pca_biplot.qnt_norm.combatRowAdj.pc",paste(pc, collapse=""),".colby_arrayrow.propprobes",prop_probes.by_var,".png"), row_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixPosition/pca_biplot.qnt_norm.combatRowAdj.pc",paste(pc, collapse=""),".colby_bmigroup.propprobes",prop_probes.by_var,".png"), bmig_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixPosition/pca_biplot.qnt_norm.combatRowAdj.pc",paste(pc, collapse=""),".colby_batch.propprobes",prop_probes.by_var,".png"), batch_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixPosition/pca_biplot.qnt_norm.combatRowAdj.pc",paste(pc, collapse=""),".colby_maternalage.propprobes",prop_probes.by_var,".png"), age_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixPosition/pca_biplot.qnt_norm.combatRowAdj.pc",paste(pc, collapse=""),".colby_stageweeks.propprobes",prop_probes.by_var,".png"), weeks_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixPosition/pca_biplot.qnt_norm.combatRowAdj.pc",paste(pc, collapse=""),".colby_trimester.propprobes",prop_probes.by_var,".png"), tri_plt, units="px", width=3000, height=3000)
  ggsave(paste0(pca_dir, "comBat_sentrixPosition/pca_biplot.qnt_norm.combatRowAdj.pc",paste(pc, collapse=""),".colby_bmi.propprobes",prop_probes.by_var,".png"), bmi_plt, units="px", width=3000, height=3000)
  
}

```

#### Save matrix and metadata with PCA

```{r}

# Save methylation matrix
write.table(minfi::getM(qnt.mSetSqFlt), paste0(res_dir,"methylation_matrix.mvals.tsv"), sep="\t", quote=FALSE, row.names=TRUE, col.names=NA)
write.table(minfi::getBeta(qnt.mSetSqFlt), paste0(res_dir,"methylation_matrix.betavals.tsv"), sep="\t", quote=FALSE, row.names=TRUE, col.names=NA)


# Save methylation matrix for func normalisation
write.table(minfi::getM(fun.mSetSqFlt), paste0(res_dir,"fun_methylation_matrix.mvals.tsv"), sep="\t", quote=FALSE, row.names=TRUE, col.names=NA)
write.table(minfi::getBeta(fun.mSetSqFlt), paste0(res_dir,"fun_methylation_matrix.betavals.tsv"), sep="\t", quote=FALSE, row.names=TRUE, col.names=NA)

# Save metadata with PCA
write.table(fun.pca_metadata, paste0(res_dir,"methylation_sample_metadata.tsv"), sep="\t", quote=FALSE, row.names=FALSE)

# Save annotation data
write.table(as.data.frame(annotdata)[,1:24], paste0(res_dir,"methylation_probe_data.tsv"), sep="\t", quote=FALSE, row.names=FALSE)

```
