---
title: "Gene expression and methylation comparison"
format: html
editor: visual
---

## Load the libraries

```{r}
library(tidyverse)
library(cowplot)
```

## Load the methylation data

```{r}
# Defined main dir
scripts_dir <- getwd()
res_dir <- paste0(scripts_dir, "/results/")

#get the gene expression data
geneExp <- readxl::read_xlsx("Gene-Expression-Data.xls") %>% as.data.frame()

# read in the mvalues for cpgs in dmrs
mvalues <- read.delim("results/dmr/dmrff_results.stage_as_continuous.treat_log2FC0.1.cpgMvals.tsv")

rownames(mvalues) <- mvalues$CpG
mvalues <- mvalues[, -1:-3]
mvalues <- mvalues[,paste0("X",geneExp$EmbryoID)]



#get the dmrs
dmrs <- read.delim("results/dmr/dmrff_results.stage_as_continuous.treat_log2FC0.1.withCpG_mValClusters_TallFormat.tsv")

#get the metadata for the samples
metadata <- read.table(paste0(res_dir,"methylation_sample_metadata.tsv"), sep="\t", header=TRUE)
rownames(metadata) <- metadata$Sample_Name

# 
metadata <- metadata[ match(geneExp$EmbryoID,metadata$Sample_Name),]


 

dmrs <- dmrs[ dmrs$SYMBOL %in% colnames(geneExp),]
topDMRs <- dmrs %>% group_by(SYMBOL) %>% slice_max(n,n=1) %>% pull(DMR_ID) %>% unique()

dmrs <- dmrs[ dmrs$DMR_ID %in% topDMRs,]

```

Model each CpG vs the gene expression

```{r}

  getStats <- function(CpG, gene,fullModel=TRUE){
    
    if(fullModel){
      fit <- lm(geneExp[,gene] ~ t(mvalues[CpG,]) + metadata$sex + metadata$mat_age + metadata$mat_bmi + metadata$Batch + metadata$Tissue..mg. + metadata$DNA..ng.ul. + metadata$Z_RRNA_ABSORB )
    } else{
          fit <- lm(geneExp[,gene] ~ t(mvalues[CpG,]))
    }

    pvalue <- summary(fit)$coefficients[2,"Pr(>|t|)"]
    r2 <- summary(fit)$r.squared
    cor <- cor(geneExp[,gene], t(mvalues[CpG,]), use="pairwise.complete.obs")[1,1]
    corstage <- cor(geneExp[,gene], metadata$Z_STAGE_WEEKS, use="pairwise.complete.obs")
    cor.pvalue <- cor.test(geneExp[,gene], t(mvalues[CpG,]))$p.val
    df <- data.frame(gene,CpG,pvalue,r2,cor,cor.pvalue)
    return(df)
  }
  
  
genes <- c(unique(dmrs$SYMBOL),"HOXA10","HOXA11")



FullRes <- lapply(genes,function(gene) {
  print(gene)
  if(gene %in% c("HOXA10","HOXA11")){
       CpGs <- dmrs[ dmrs$SYMBOL=="HOXA3","CpG"]
       dmrID <- dmrs[ dmrs$SYMBOL=="HOXA3","DMR_ID"]
  } else{
      CpGs <- dmrs[ dmrs$SYMBOL==gene,"CpG"]
      dmrID <- dmrs[ dmrs$SYMBOL==gene,"DMR_ID"]
  } 
  
  res <- bind_rows(lapply(CpGs, getStats, gene))
  res$dmrID <- dmrID

  return(res)
  }) %>% bind_rows()




write.table(FullRes, "geneExpvsMeth.txt", sep="\t", quote=FALSE, row.names=FALSE)

plotGeneCpg <- function(CpG,geneExp,metadata){
  
  dat <- bind_cols(geneExp[,c("Stage (pcw)",gene)],t(mvalues[CpG,]))
  colnames(dat)[1] <- "stage"
  pivot_longer(dat, cols = -stage, names_to = "variable", values_to = "value") %>%
  ggplot(aes(x=stage,y=value,color=variable)) + 
    geom_point() +
    facet_wrap(~variable,scales = "free") +
    cowplot::theme_cowplot()

}





```
